File: leetcode/python/1155_number_of_dice_rolls_with_target_sum.py
class Solution:
    def numRollsToTarget(self, n: int, k: int, target: int) -> int:
        MAX = 10 ** 9 + 7
        dp = [0] * (target + 1)
        for j in range(1, min(k+1, target+1)):
            dp[j] = 1

        for _ in range(1, n):
            for j in range(target, 0, -1):
                dp[j] = sum(dp[max(j-k, 0):j]) % MAX
        return dp[-1]

s = Solution()
print(s.numRollsToTarget(n = 1, k = 6, target = 3))
print(s.numRollsToTarget(n = 2, k = 6, target = 7))
print(s.numRollsToTarget(n = 30, k = 30, target = 500))

File: leetcode/rust/src/1155_number_of_dice_rolls_with_target_sum.rs
impl Solution {
    pub fn num_rolls_to_target(n: i32, k: i32, target: i32) -> i32 {
        const MOD: i32 = 1_000_000_007;
        let mut dp = vec![0; target as usize + 1];
        
        for i in 1..=k.min(target) as usize {
            dp[i] = 1;
        }

        for _ in 1..n as usize {
            for i in (1..=target as usize).rev() {
                dp[i] = 0;
                for j in 1..=k.min(i as i32) as usize {
                    dp[i] = (dp[i] + dp[i - j]) % MOD;
                }
            }
        }

        dp.last().unwrap().clone()
    }
}